bazel_tools/proto.bzl

```diff
@@ -37,7 +43,7 @@ def get_plugin_runfiles(tool, plugin_runfiles):
 
 def _proto_gen_impl(ctx):
     src_descs = [src[ProtoInfo].direct_descriptor_set for src in ctx.attr.srcs]
-    dep_descs = [dep[ProtoInfo].direct_descriptor_set for dep in ctx.attr.deps]
+    dep_descs = [depset for dep in ctx.attr.deps for depset in dep[ProtoInfo].transitive_descriptor_sets.to_list()]
```

Comment:
It looks like this changed from using only direct deps to using transitive deps. Why do we need this change?

Reply:
Otherwise, when compiling, for example, _transaction.proto_, it pulls in _value.proto_ but not its dependencies (e.g. _empty.proto_), so it refuses to compile. This wasn't a problem before because we only used `proto_gen` without transitive Protobuf dependencies (AFAIK), so I believe it was an oversight.

I might be wrong.

Reply:
Sounds sensible, thanks for the explanation.

Reply:
Looking at the use-cases of `proto_gen` I think the `deps` attribute in its current form could be removed altogether. Instead, we can use `.transitive_descriptor_sets` on `srcs`. These `deps` are already listed on the corresponding `proto_library` targets, so there is no need to repeat them on `proto_gen`. 

 [`go_proto_library`](https://github.com/bazelbuild/rules_go/blob/983d48f9d2695a0a9d898bd85dde90d410629b1b/proto/core.rst#go-proto-library) follows this pattern. It uses `protos` in place of `srcs` and `deps` is used to provide Go dependencies to the generated code.

Reply:
This seems to work:

```diff
diff --git a/bazel_tools/proto.bzl b/bazel_tools/proto.bzl
index 94d696935..4791fdfe9 100644
--- a/bazel_tools/proto.bzl
+++ b/bazel_tools/proto.bzl
@@ -44,9 +44,7 @@ def get_plugin_runfiles(tool, plugin_runfiles):
     return files
 
 def _proto_gen_impl(ctx):
-    src_descs = [src[ProtoInfo].direct_descriptor_set for src in ctx.attr.srcs]
-    dep_descs = [depset for dep in ctx.attr.deps for depset in dep[ProtoInfo].transitive_descriptor_sets.to_list()]
-    descriptors = src_descs + dep_descs
+    descriptors = [depset for src in ctx.attr.srcs for depset in src[ProtoInfo].transitive_descriptor_sets.to_list()]
 
     sources_out = ctx.actions.declare_directory(ctx.attr.name + "-sources")
```

Is that what you meant, or is there further simplification we can do here?

Reply:
Yes, that's what I meant. The `deps` attribute can then also be removed from the rule definition. I.e.
``` diff
diff --git a/bazel_tools/proto.bzl b/bazel_tools/proto.bzl
index 94d696935b9..74cfdb142e5 100644
--- a/bazel_tools/proto.bzl
+++ b/bazel_tools/proto.bzl
@@ -133,8 +133,7 @@ def _proto_gen_impl(ctx):
 proto_gen = rule(
     implementation = _proto_gen_impl,
     attrs = {
-        "srcs": attr.label_list(allow_files = True),
-        "deps": attr.label_list(providers = [ProtoInfo]),
+        "srcs": attr.label_list(providers = [ProtoInfo]),
         "plugin_name": attr.string(),
         "plugin_exec": attr.label(
             cfg = "host",
```
(I noticed that the definition of `srcs` is slightly off. It claims to accept files but the implementation expects a `ProtoInfo` provider)

Reply:
Sounds good. Here we go: https://github.com/digital-asset/daml/pull/8120